enhancements/storage/vsphere-problem-detector.md

```diff
@@ -0,0 +1,230 @@
+* Storage Distributed Resource Scheduler (DRS) is not supported.
+* vMotion (i.e. VM migration) is not supported.
+
+Such support cases are often frustrating, as there is high variety of vSphere environments, error messages we get from vSphere are not always useful and at the same time users either can't use the cluster (can't create / attach PVs) or even lose data.
+
+Secondary objective is that we (as OCP engineering / product management / support) do not know how our customers configure vSphere / OCP on top of it and if a future OCP release may break such OCP installation.
+
+* We'd like to know HW version of virtual machines they use (necessary for vSphere CSI driver & removing in-tree cloud provider).
+* We'd like to know what vSphere version they use (same as above).
+
+### Goals
+
+* User can detect and fix OCP / vSphere storage related configuration issues by themselves.
+* Our support can easily check OCP was correctly configured and supportable.
+* We get metrics about vSphere environments where OCP runs.
+* Add pre-flight checks to installer to fail installation if provided vSphere cluster / credentials are not suitable.
+
+### Non-Goals
+
+* Any automatic fix of the environment.
+
+## Proposal
+
+Run a dedicated vsphere-problem-detector operator that periodically tests OCP / vSphere:
+
+* The operator reports metrics, some of them sent via telemetry to us.
+* The operator reports alerts for common configuration issues.
+* The operator logs performed checks + more detailed info about OCP/vSphere config, so it's available in must-gather.
+
+Add pre-flight checks to installer.
+
+### Checks
+
+These checks will be performed regularly and / or in the installer:
+
+| Test | Periodic | Installer |
+|-|-|-|
+| OCP has enough permissions (e.g. list recent tasks and list datastores). | yes | yes |
+| Kubelets run with cloud-provider configured (i.e. have `Node.Spec.ProviderID`). |  yes | no |
+| All VMs have `disk.enableUUID=TRUE`. |  yes | no |
+| Datastore name used in OCP config is not too long. |  yes | yes |
+| Datastore names used in StorageClasses and PVs are not too long. |  yes | no |
+| Stretch goal: OCP can create a volume + attach it to a VM (this usually works when the checks above pass). |  yes | no |
+| Stretch goal: Was a VM / storage recently migrated using vMotion? We don't support this (yet). | yes | no |
+| Stretch goal: Is Storage Distributed Resource Scheduler enabled? We don't support this. | yes | no |
+
+When a previous test failed, the initial test frequency is once per 1 hour, doubling with every failed attempt up to once per 8 hours max.
+When a previous test succeeded, the frequency is once per 8 hours.
+
+### Metrics
+
+We want these metrics to be transferred via telemetry to us:
+
+* List of failed periodic tests.
+* HW version of vSphere VMs.
+* vCenter version.
+* ESXi host version (of each host).
+* List of installed storage plugins (3rd party vendor VIBs) - if possible.
```

Comment:
We should be able to get vibs, assuming the user would have rights to query that.

```javascript
hostLength = clusterComputeResource.host.length

for (i =0; i < hostLength; i++) {
    System.log(clusterComputeResource.host[i].summary.config.product.version)
    System.log(clusterComputeResource.host[i].summary.config.product.build)

    softwarePackages = clusterComputeResource.host[i].configManager.imageConfigManager.fetchSoftwarePackages()

    System.log(softwarePackages.length)

    for(j=0;j<softwarePackages.length;j++){
        System.log(softwarePackages[j].name)
    }
}
```